Reusable blocks: Improve UX for non-privileged users

noisysocks opened the pull request:
### What this is

Fixes #12338.

Improves the UX of creating, editing, and deleting a reusable block when logged in as an author or contributor by disabling the _Add to Reusable Blocks_, _Edit_, and _Remove from Reusable Blocks_ buttons when necessary.

This is accomplished under the hood by introducing the `canUser()` selector to `core-data` which allows callers to query whether the REST API supports performing a given action on a given resource, e.g. one can query whether the logged in user can create posts by running `wp.data.select( 'core' ).canUser( 'create', 'posts' )`.

The existing `hasUploadPermissions()` selector is changed to use `canUser( 'create', 'media' )` under the hood.

### How it looks

_Add to Reusable Blocks_ button is visible for admins and editors:

![admin - add](https://user-images.githubusercontent.com/612155/50619911-71a5c380-0f50-11e9-8f30-acf8304b81e6.png)

_Add to Reusable Blocks_ button is hidden for contributors:

![contributor - add](https://user-images.githubusercontent.com/612155/50619915-78343b00-0f50-11e9-84c2-03dbee5f21b9.png)

_Edit_ button is enabled and _Remove from Reusable Blocks_ is visible for admins and editors:

![admin - edit remove](https://user-images.githubusercontent.com/612155/50619935-9601a000-0f50-11e9-8a6f-a679c9da4155.png)

_Edit_ button is disabled and _Remove from Reusable Blocks_ is hidden for authors:

![author - edit remove](https://user-images.githubusercontent.com/612155/50619918-86825700-0f50-11e9-969f-1c139bc113a4.png)

### How to test

1. Try to create a reusable block as a contributor. The _Add to Reusable Blocks_ button should be disabled.
1. Try to updating another user's reusable block as an author. The _Edit_ button should be disabled.
1. Try to deleting another user's reusable block as an author or contributor. The _Remove from Reusable Blocks_ button should be disabled.

noisysocks committed:
Improve reusable block UX for non-privileged users

Improves the UX of creating, editing, and deleting a reusable block when
logged in as an author or contributor by disabling the _Add to Reusable
Blocks_, _Edit_, and _Remove from Reusable Blocks_ buttons when
necessary.

This is accomplished under the hood by introducing the `canUser()`
selector to `core-data` which allows callers to query whether the REST
API supports performing a given action on a given resource, e.g. one can
query whether the logged in user can create posts by running
`wp.data.select( 'core' ).canUser( 'create', 'posts' )`.

The existing `hasUploadPermissions()` selector is changed to use
`canUser( 'create', 'media' )` under the hood.

noisysocks commented:
This should be ready!

@mtias: Reckon we could plop this back into 4.7? I'd like to have this fixed for 5.0.1 since it's a pretty awful UX otherwise. Not sure where we are timing wise.

talldan commented on packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js:
```diff
@@ -29,6 +30,7 @@ export function ReusableBlockConvertButton( {
 				<MenuItem
 					className="editor-block-settings-menu__control"
 					icon="controls-repeat"
+					disabled={ ! canCreateBlocks }
```
I think this is the only block settings menu item that ends up disabled. Most of the others are just not rendered when they're not valid in a specific context and I think it's worth following the same pattern here. The same pattern (hiding instead of disabling) is also used for other bits of UI when users don't have the capability.

noisysocks replied:
> The same pattern (hiding instead of disabling) is also used for other bits of UI when users don't have the capability.

Makes sense to be consistent here then 🙂 — have updated the behaviour in 708a50484fc99537a5f3368b4dbb66e134198741.

talldan commented on packages/core-data/src/selectors.js:
```diff
@@ -178,5 +178,22 @@ export function isPreviewEmbedFallback( state, url ) {
+ *
+ * @return {boolean} Whether or not the user can perform the action.
+ */
+export function canUser( state, action, resource, id ) {
```
I like the expressiveness of the canUser selector 😄

noisysocks replied:
Props to @gziolo for suggesting the API 🙂

talldan commented on packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js:
```diff
@@ -35,9 +35,16 @@ export default compose( [
 		return {
-			reusableBlock: block && isReusableBlock( block ) ? getReusableBlock( block.attributes.ref ) : null,
+			id,
+			isDisabled: !! id && (
```
I think same comment here regarding disabling vs. hiding.

talldan replied:
Also noting that in testing I didn't see the flash of this switching from enabled to disabled. That seems to be because the GET request for the block has already been made and cached. If an OPTIONS request was made instead of a GET request the issue would resurface.

noisysocks replied:
Have changed this to remove the button instead of disabling in 708a50484fc99537a5f3368b4dbb66e134198741.

We can't preload these requests because we don't know the reusable block ID ahead of time.

talldan replied:
Yep, I do understand that we can't preload. My worry is that it's only by coincidence that the user's capability is known ahead of time. If the underlying code switches to use OPTIONS instead of GET, users would start seeing this button initially be clickable but then suddenly disappear.

This is probably a nitpick though and not a blocker. I'm not even sure what the best UX would be.

noisysocks replied:
Ah, got it! I wonder if defaulting to disabled would be a better experience...

I think that, further down the track, `core-data` should modify the `canUser` state whenever it receives an `Allow` header from _any_ response, and not just when the `canUser()` resolver is called.

talldan replied:
Yeah, that's a nice idea. Not sure how it'd work, but a nice idea 😄

talldan commented on packages/core-data/src/resolvers.js:
```diff
@@ -103,7 +103,46 @@ export function* getEmbedPreview( url ) {
+			path,
+			// Ideally this would always be an OPTIONS request, but unfortunately there's
+			// a bug in the REST API which causes the Allow header to not be sent on
+			// OPTIONS requests to /posts/:id routes.
```
Is that just for /post/:id of for the all resources? I mentioned that because the logic below happens for all resources.

Is there a ticket/issue covering the bug?

noisysocks replied:
I've only checked posts but I suspect the bug affects all REST endpoints. Will do some investigating and open up a Trac ticket 👍

gziolo replied:
If this is the case, it should be fixed /cc @danielbachhuber @kadamwhite

TimothyBJacobs replied:
If you just want the `Allow` header, what about doing a `HEAD` request instead of `GET`.

danielbachhuber replied:
Looking into this further:

1. The `Allow` header is generated by `rest_send_allow_header()`, which is called on `rest_post_dispatch`.
    * https://github.com/WordPress/WordPress/blob/d2782aabc51f387c405eacc2ddbd38dfdacd4405/wp-includes/rest-api.php#L628
    * https://github.com/WordPress/WordPress/blob/d2782aabc51f387c405eacc2ddbd38dfdacd4405/wp-includes/rest-api.php#L167
2. `rest_post_dispatch` is called for every response (I think):
    * https://github.com/WordPress/WordPress/blob/47d32decd6668b3ccf100acc8a40aa0a9fc35003/wp-includes/rest-api/class-wp-rest-server.php#L352

It seems like `Allow` should be present for `OPTIONS`. It's an officially supported header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS

Can you provide steps to reproduce?

noisysocks replied:
I briefly debugged this when I noticed the issue and from memory it was something to do with `rest_handle_options_request` not calling `$request->set_url_params()` unlike `dispatch` which correctly parses and sets URL parameters.

> Can you provide steps to reproduce?

1. Create a reusable block
1. Perform a `GET http://local.wordpress.test/wp-json/wp/v2/blocks` request and note the ID of the reusable block
1. Perform a `OPTIONS http://local.wordpress.test/wp-json/wp/v2/blocks/27` request, replacing `27` with the ID from step 2. Note that an `Allow` header is not sent back
1. Perform a `GET http://local.wordpress.test/wp-json/wp/v2/blocks/27` request, replacing `27` with the ID from step 2. Note that an `Allow` header is sent back

noisysocks replied:
I created a ticket here: https://core.trac.wordpress.org/ticket/45753

noisysocks replied:
Added a link to the Trac ticket in cf2d7012f.

talldan reviewed (dismissed):
I like the canUser function, it's a nice way to express that functionality.

I think that flash of the disabled state of the block settings menu item is something that needs to be fixed. 

Preloading could work, but I think checking the resolution state is also possible and worth doing as well, since it shouldn't just depend on preloading. I think this is what embeds do:
https://github.com/WordPress/gutenberg/blob/master/packages/core-data/src/selectors.js#L41

noisysocks commented on packages/core-data/src/resolvers.js:
```diff
@@ -103,7 +103,46 @@ export function* getEmbedPreview( url ) {
  * Requests Upload Permissions from the REST API.
  */
 export function* hasUploadPermissions() {
```
Should I mark this as deprecated? I'm unsure on what the policy is post 5.0 🙂

talldan replied:
I don't know either. Might be worth pinging someone on slack. Will hold off approving until the answer is known.

noisysocks replied:
cc. @youknowriad

noisysocks replied:
It's probably fine leaving this as is where we support both `canUser` and `hasUploadPermissions`. We can always deprecate later.

youknowriad replied:
Yes, for now, we should leave those, we can add a deprecation message if needed but without specifying a version. 

This is something we should discuss more in the next #core-js chats.

noisysocks replied:
I noted this in [the agenda for the next meeting](url
https://docs.google.com/document/d/1QADtwgPif4-fCKIlS6rPP9CXhGgWZI2wFx8UfcXf7kk/edit).

talldan replied:
@noisysocks I noticed you `@deprecated` the selector, but the resolver hasn't been deprecated. I guess technically they're the same interface and the resolver wouldn't be called directly, but I wanted to double-check if it was intentional.

noisysocks replied:
That's a good point. I think selectors are the public interface here but it wouldn't hurt to have the `@deprecated` attribute in the resolver as well.

noisysocks replied:
Done in 4ea46a5d6c211e9614e9d2304cf1563a73394fc3!

noisysocks committed:
Expand canUser() selector documentation

noisysocks committed:
Preload OPTIONS request that determins if user can create a block

noisysocks commented on packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js:
```diff
@@ -80,6 +83,7 @@ export default compose( [
 				! isReusableBlock( blocks[ 0 ] ) ||
 				! getReusableBlock( blocks[ 0 ].attributes.ref )
 			),
+			canCreateBlocks: canUser( 'create', 'blocks' ),
```
👌 I've added this OPTIONS request to the list of requests we preload in 0115ae6a7.

Noting that this is a change we'll need to make to core after this PR is merged.

noisysocks replied:
👌 I've added this OPTIONS request to the list of requests we preload in bc0709f27cfa4fef3b30847c7aa328a69850799f.

Noting that this is a change we'll need to make to core after this PR is merged.

noisysocks committed:
Hide 'Add to Reusable Blocks' and 'Remove from Reusable Block' buttons

Hide the 'Add to Reusable Blocks' and 'Remove from Reusable Blocks'
buttons instead of disabling them when the user does not have the
necessary permissions.

noisysocks commented on packages/core-data/src/selectors.js:
```diff
@@ -178,5 +178,22 @@ export function isPreviewEmbedFallback( state, url ) {
+}
+
+/**
+ * Returns whether the current user can perform the given action on the given
```
👌 537661450ac04c7d4867c424887dd64fd225b672

noisysocks commented:
@gziolo any thoughts on this one?

gziolo commented on packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js:
```diff
@@ -35,18 +35,27 @@ export default compose( [
 		};
 	} ),
-	withDispatch( ( dispatch, { onToggle = noop } ) => {
+	withDispatch( ( dispatch, { clientId, onToggle = noop }, { select } ) => {
```
`select` - that's interesting to see it have much broader application than I anticipated. I hope there are no drawbacks of that approach :)

noisysocks replied:
I like that in this case it lets us avoid passing unnecessary props to the component.

gziolo commented on packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js:
```diff
@@ -54,32 +54,39 @@ export default compose( [
-				! isReusableBlock( blocks[ 0 ] ) ||
-				! getReusableBlock( blocks[ 0 ].attributes.ref )
+			// Show 'Convert to Regular Block' when selected block is a reusable block
+			isVisible: isReusable || (
```
It's probably better to move `isVisible` outside of the object - the way it was before. It makes return value harder follow at the moment.

noisysocks replied:
👍 1c3f722b5

gziolo commented on packages/core-data/src/selectors.js:
```diff
@@ -178,5 +178,27 @@ export function isPreviewEmbedFallback( state, url ) {
+ * @param {string}  resource REST resource to check, e.g. 'media' or 'posts'.
+ * @param {?string} id       ID of the rest resource to check.
+ *
+ * @return {boolean} Whether or not the user can perform the action.
```
Can we add the information about the default behavior when it is not found in the store?

noisysocks replied:
👍 Done as part of 62bcb1903606588ca2586bba5cf981843b459261.

gziolo commented:
> The existing `hasUploadPermissions()` selector is changed to use `canUser( 'create', 'media' )` under the hood.

@imath - I think you implemented the original version of this behavior when working on media upload permissions. Can you double check if the generalized version is still correct?

@danielbachhuber or @kadamwhite - it would be nice to get your review on the bits that interact with REST API.

danielbachhuber commented:
> Improves the UX of creating, editing, and deleting a reusable block when logged in as an author or contributor by disabling the _Add to Reusable Blocks_, _Edit_, and _Remove from Reusable Blocks_ buttons when necessary.

Does this change the UI at all? If so, can you include screenshots and/or GIFs in the description?

danielbachhuber commented:
Also:
> The only important thing from my perspective is to ensure that `canUser` has the best strategy for returning the default value.

At first glance, I'm not sure how well this `canUser` abstraction is going to work in practice. WordPress' roles and capabilities system is quite nuanced, and abstractions on top of it can be problematic. Just want to flag that.

imath commented:
@gziolo Hi,

I've just tested the PR. I confirm logged in as an Administrator I'm still able to upload files and that being logged in as a contributor I can only use external URLs to add media. So it seems to be working as expected for this part 👍 

@noisysocks FYI I also tested the Reusable blocks UI changes being logged in as a contributor: 
- I confirm I cannot create new reusable block.
- I can use another user reusable block but i cannot edit it.
- I can access to the `/wp-admin/edit.php?post_type=wp_block` manage screen but can only export the blocks created by others as json.

tofumatt commented on packages/core-data/src/resolvers.js:
```diff
@@ -116,5 +155,7 @@ export function* hasUploadPermissions() {
 	}
 
-	yield receiveUploadPermissions( includes( allowHeader, 'POST' ) );
+	const key = compact( [ action, resource, id ] ).join( '/' );
```
You'd need template strings for the first part:

```suggestion
	const path = `/wp/v2/${ resource }` + ( id ? `/${ id }` : '' );
```

I think it's a bit easier to sort out though, yeah.

noisysocks replied:
Was this suggestion meant to go in this thread? https://github.com/WordPress/gutenberg/pull/12378#discussion_r241679744

tofumatt commented on packages/core-data/src/resolvers.js:
```diff
@@ -116,5 +155,7 @@ export function* hasUploadPermissions() {
-	yield receiveUploadPermissions( includes( allowHeader, 'POST' ) );
+	const key = compact( [ action, resource, id ] ).join( '/' );
+	const isAllowed = includes( allowHeader, method );
+	yield receiveUserPermissions( key, isAllowed );
```
At the very least could we add the ticket URL in the comment so it can be tracked/checked on?

noisysocks replied:
> Should it be named receiveUserPermission using singular version to reflect that you can provide only one permission at the time?

👍 02d9502cd

> At the very least could we add the ticket URL in the comment so it can be tracked/checked on?

I think this is referring to this thread? https://github.com/WordPress/gutenberg/pull/12378#discussion_r239331728 If so I added a link to the Trac ticket in 
https://github.com/WordPress/gutenberg/commit/cf2d7012f039500091aae18ce23e5698f77fa2e0.

tofumatt commented on packages/core-data/src/resolvers.js:
```diff
@@ -103,7 +103,46 @@ export function* getEmbedPreview( url ) {
+			parse: false,
+		} );
+	} catch ( error ) {
+		return;
```
😟 This seems like it should throw. Why are we catching an ignoring the error without doing anything with it?

noisysocks replied:
`error` here refers to the API error that resulted from our OPTIONS request. Doing nothing is correct in this case as we want to leave the store as is. I've added a comment to explain this in e0a7269a1518e89efef5fbd294af9c7417e35c49.

tofumatt commented on packages/core-data/src/test/selectors.js:
```diff
@@ -117,3 +118,30 @@ describe( 'isPreviewEmbedFallback()', () => {
+		const state = deepFreeze( {
+			userPermissions: {},
+		} );
+		expect( canUser( state, 'create', 'media' ) ).toBe( true );
```
Yeah, `true` as the default seems wrong to me. I'd rather UI update after API requests and have extra settings added than them removed.

noisysocks replied:
I've changed `canUser()` to return `false` by default in 62bcb1903606588ca2586bba5cf981843b459261.

tofumatt commented on packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js:
```diff
@@ -54,32 +54,39 @@ export default compose( [
+			blocks.length === 1 &&
+			blocks[ 0 ] &&
+			isReusableBlock( blocks[ 0 ] ) &&
+			!! getReusableBlock( blocks[ 0 ].attributes.ref )
```
Any reason to use `!!` here? I just always find it weird to see 🤷‍♂️

noisysocks replied:
It's to ensure that `typeof isReusable === 'boolean'`.

tofumatt commented on packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js:
```diff
@@ -54,32 +54,39 @@ export default compose( [
+
+				every( blocks, ( block ) => (
+					// Guard against the case where a regular block has *just* been converted
+					!! block &&
```
Again, any reason for the `!!`?

noisysocks replied:
It's to ensure that `typeof isVisible === 'boolean'`.

tofumatt reviewed (dismissed):
This is looking good—I like the `canUser` selector API, but I think there's some code cleanup here to do and I have two concerns:

1. why is `true` the default response for `canUser()`?
2. Why are we ignoring the API request throwing an error?

noisysocks committed:
Link to the Trac ticket for the REST API bug that we're working around

aduth commented on docs/designers-developers/developers/data/data-core.md:
```diff
@@ -148,6 +148,28 @@ Return Upload Permissions.
+
+ * state: Data state.
+ * action: Action to check. One of: 'create', 'read', 'update',
+                          'delete'.
```
> This indentation is kinda odd; I'm not sure why this is indented where it is...

Likely because it's unprocessed from the space-aligned JSDoc. Could probably be improved with some additional processing (`lines.map( trim ).join( ' ' )`?). It has no difference in the Markdown preview though, since the whitespace is collapsed.

aduth commented on packages/block-library/src/block/edit.js:
```diff
@@ -159,6 +162,7 @@ export default compose( [
 			isFetching: isFetchingReusableBlock( ref ),
 			isSaving: isSavingReusableBlock( ref ),
 			block: reusableBlock ? getBlock( reusableBlock.clientId ) : null,
+			canUpdateBlock: !! reusableBlock && ! reusableBlock.isTemporary && canUser( 'update', 'blocks', ref ),
```
> Any reason you're checking `!! reusableBlock` rather than just using `reusableBlock`? Just curious!

If a value is falsey in this scenario, the specific value is returned in the left-hand of the condition.

If `reusableBlock` was `undefined` for example, without the `!!`, `canUpdateBlock` would be passed as `undefined`, despite its name indicating that it would be a boolean. As written, it satisfies an implicit contract of providing a boolean value. Practically speaking this would rarely have much an impact, except in a scenario where someone expected it to be the particular boolean form (`=== false`).

noisysocks committed:
End error messages with a period

noisysocks commented on packages/core-data/src/resolvers.js:
```diff
@@ -103,7 +103,46 @@ export function* getEmbedPreview( url ) {
+		throw new Error( `'${ action }' is not a valid action` );
+	}
+
+	const path = id ? `/wp/v2/${ resource }/${ id }` : `/wp/v2/${ resource }`;
```
I like that with the more verbose version that you can glance at the code and see the "shape" of the URL.

noisysocks committed:
s/receiveUserPermissions/receiveUserPermission

noisysocks commented on packages/core-data/src/selectors.js:
```diff
@@ -178,5 +178,27 @@ export function isPreviewEmbedFallback( state, url ) {
  */
 export function hasUploadPermissions( state ) {
-	return state.hasUploadPermissions;
+	return canUser( state, 'create', 'media' );
```
See this thread: https://github.com/WordPress/gutenberg/pull/12378#discussion_r239336329. I think it's fine to leave for now and come back when we have a policy on how to deprecate methods post 5.0.

noisysocks replied:
I added a `@deprecated` section to the JSDoc block as part of 62bcb1903606588ca2586bba5cf981843b459261.

noisysocks committed:
Have canUser() return false by default

noisysocks committed:
Move isVisible logic into its own variable

noisysocks committed:
Add comment explaining why we ignore API errors

noisysocks commented on packages/core-data/src/resolvers.js:
```diff
@@ -103,7 +103,46 @@ export function* getEmbedPreview( url ) {
+
+	const method = methods[ action ];
+	if ( ! method ) {
+		throw new Error( `'${ action }' is not a valid action` );
```
👍 ccf4936cc

noisysocks commented:
Thanks all for the feedback! I think that I've addressed all outstanding comments.

---

> @noisysocks - nice work on the PR. I personally think it is quite close to be considered ready. My comments are mostly nitpicks. The only important thing from my perspective is to ensure that canUser has the best strategy for returning the default value. It feels like assuming there is no permission is safer.

In 62bcb19 I changed `canUser()` to return `false` by default and added an optional `defaultIsAllowed` param which lets the caller decide what the default return value should be.

> Does this change the UI at all? If so, can you include screenshots and/or GIFs in the description?

Yes. I've added some screenshots to the description.

> At first glance, I'm not sure how well this canUser abstraction is going to work in practice. WordPress' roles and capabilities system is quite nuanced, and abstractions on top of it can be problematic. Just want to flag that.

Could you elaborate on this?

There might be some confusion here. `canUser()` checks a user's ability to perform a particular REST API action on a particular REST API resource. For example, `canUser( 'create' 'media' )` performs an `OPTIONS /wp/v2/media` requests and checks that `POST` appears in the resultant `Allow` header. It **does not** check that a user or a role has a particular capability. 

> 1. why is true the default response for canUser()?
> 2. Why are we ignoring the API request throwing an error?

I've changed the default response of `canUser()` in 62bcb19 and have explained why we ignore the API error in https://github.com/WordPress/gutenberg/pull/12378#discussion_r244897815.

noisysocks committed:
Typo: s/4.8/4.9/

noisysocks commented:
I think this is ready. Is anyone able to give it a last look over?

noisysocks committed:
Add @deprecated comment to hasUploadPermissions() resolver

noisysocks commented:
Pinging @WordPress/gutenberg-core for a review. Would like to get this in. The current experience in WordPress 5.0. is pretty horrid 🙂

noisysocks commented:
In 91e49659aba3c0ee947e65dad596579df5c81263 I've used `deprecated()` to mark `hasUploadPermission()` as deprecated as per [the discussion in Slack](https://wordpress.slack.com/archives/C5UNMSU4R/p1547023091120300).

noisysocks committed:
Use deprecated() to mark hasUploadPermissions() as deprecated

talldan commented on packages/editor/src/components/media-placeholder/index.js:
```diff
@@ -258,10 +258,10 @@ export class MediaPlaceholder extends Component {
 
 	return {
-		hasUploadPermissions: hasUploadPermissions(),
+		hasUploadPermissions: canUser( 'create', 'media', undefined, true ),
```
This is a personal preference, but I'm not such a big fan of having to pass undefined as an argument here. Some alternatives (but don't feel like it's something that needs to be changed):
```
defaultTo( canUser( 'create', 'media' ), true );
```

```
canUser( [ 'create', 'media' ], true );
```

```
canUser( { action: 'create', resource: 'media', default: true } );
```

talldan reviewed (approved):
Re-tested both media and reusable blocks as a contributor. Works well. 

A++++ great PR

noisysocks committed:
Typo: s/ot/to/

noisysocks committed:
Have `canUser()` return `undefined` by default

talldan commented:
IIRC this one also needed a change in core to preload that request.

noisysocks committed:
Update version: s/4.9/5.0/

noisysocks commented:
Getting this in early so that there's time for it to bake in 5.0.

> IIRC this one also needed a change in core to preload that request.

Thanks for the reminder. I wonder what the best way to do this is... Do we check how the plugin PHP files differ when updating Core's script dependencies? (e.g. `git diff v4.9 v5.0 **.php`) If so, we'll catch this change. If not, perhaps opening a ticket is the safest bet.

talldan commented:
@noisysocks Not sure, are the files easily diffable, they look quite different. This one was done as a trac ticket, so maybe that's the best option:
https://core.trac.wordpress.org/changeset/43833

aduth commented on packages/core-data/package.json:
```diff
@@ -24,6 +24,7 @@
 		"@babel/runtime": "^7.0.0",
 		"@wordpress/api-fetch": "file:../api-fetch",
 		"@wordpress/data": "file:../data",
+		"@wordpress/deprecated": "file:../deprecated",
```
There should have been a corresponding change to the root `package-lock.json`. I'm not sure how this passed the build task which checks for local uncommitted changes, but it shouldn't have.

noisysocks replied:
Huh! Looks like `check-local-changes` doesn't pick this up because `precheck-local-changes` doesn't run `npm install`.

Will fix this up in a seperate PR.

noisysocks replied:
https://github.com/WordPress/gutenberg/pull/13462

paaljoachim commented:
A test that I made:
Contributor made a new post. Submitted for Review.
Logged in as admin. Went to the newly created contributor post and took it over. Made a reusable block. Clicked to save as pending.
Logged in as contributor. Went to post. Noticed that contributor could delete the reusable block.

talldan commented:
@paaljoachim What version are you testing against? This has only just been released in the plugin today afaik.

I couldn't reproduce testing against master.

paaljoachim commented:
Hey Daniel @talldan

Lets just call it an error on my end. I made the following comment:
https://make.wordpress.org/test/2019/02/05/call-for-testing-gutenberg-5-0/#comment-1170
